File: goals_service/app/schemas/goal.py
from pydantic import BaseModel, Field, validator, field_validator
from typing import Optional, List
from datetime import datetime, date
from uuid import UUID
from app.models.goal import GoalType, GoalStatus, GoalPriority


class GoalBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=1000)
    goal_type: GoalType
    priority: GoalPriority = GoalPriority.MEDIUM
    target_amount: float = Field(..., gt=0)
    currency: str = Field(default="USD", max_length=3)
    target_date: Optional[datetime] = None
    is_milestone_based: bool = False

    @field_validator('target_date', mode='before')
    @classmethod
    def parse_target_date(cls, v):
        """Accept both date and datetime formats"""
        if v is None:
            return v
        if isinstance(v, datetime):
            return v
        if isinstance(v, date):
            return datetime.combine(v, datetime.min.time())
        if isinstance(v, str):
            # Try date-only format first (YYYY-MM-DD)
            if len(v) == 10 and '-' in v:
                try:
                    return datetime.strptime(v, '%Y-%m-%d')
                except ValueError:
                    pass
            # Try datetime format
            for fmt in ['%Y-%m-%dT%H:%M:%S', '%Y-%m-%d %H:%M:%S', '%Y-%m-%dT%H:%M:%S.%f']:
                try:
                    return datetime.strptime(v, fmt)
                except ValueError:
                    continue
        return v  # Let Pydantic handle validation error

    @validator('target_amount')
    def validate_target_amount(cls, v):
        if v <= 0:
            raise ValueError('Target amount must be greater than 0')
        return round(v, 2)


class GoalCreate(GoalBase):
    pass


class GoalUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=1000)
    goal_type: Optional[GoalType] = None
    priority: Optional[GoalPriority] = None
    target_amount: Optional[float] = Field(None, gt=0)
    currency: Optional[str] = Field(None, max_length=3)
    target_date: Optional[datetime] = None
    status: Optional[GoalStatus] = None
    is_milestone_based: Optional[bool] = None

    @field_validator('target_date', mode='before')
    @classmethod
    def parse_target_date(cls, v):
        """Accept both date and datetime formats"""
        if v is None:
            return v
        if isinstance(v, datetime):
            return v
        if isinstance(v, date):
            return datetime.combine(v, datetime.min.time())
        if isinstance(v, str):
            if len(v) == 10 and '-' in v:
                try:
                    return datetime.strptime(v, '%Y-%m-%d')
                except ValueError:
                    pass
            for fmt in ['%Y-%m-%dT%H:%M:%S', '%Y-%m-%d %H:%M:%S', '%Y-%m-%dT%H:%M:%S.%f']:
                try:
                    return datetime.strptime(v, fmt)
                except ValueError:
                    continue
        return v

    @validator('target_amount')
    def validate_target_amount(cls, v):
        if v is not None and v <= 0:
            raise ValueError('Target amount must be greater than 0')
        return round(v, 2) if v is not None else v


class GoalProgressUpdate(BaseModel):
    current_amount: float = Field(..., ge=0)

    @validator('current_amount')
    def validate_current_amount(cls, v):
        return round(v, 2)


class GoalResponse(GoalBase):
    id: int
    user_id: int
    workspace_id: UUID
    current_amount: float
    progress_percentage: float
    status: GoalStatus
    start_date: datetime
    created_at: datetime
    updated_at: Optional[datetime] = None
    is_read_only: bool = False

    class Config:
        from_attributes = True


class GoalWithMilestones(GoalResponse):
    milestones: List['MilestoneResponse'] = []


class GoalListResponse(BaseModel):
    items: List[GoalResponse]
    total: int
    page: int
    size: int
    pages: int


class GoalStatistics(BaseModel):
    total_goals: int
    active_goals: int
    completed_goals: int
    total_target_amount: float
    total_current_amount: float
    overall_progress: float
    currency: str
    goals_by_type: dict
    goals_by_priority: dict


# Milestone schemas
class MilestoneBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=500)
    target_amount: float = Field(..., gt=0)
    order_index: int = 0

    @validator('target_amount')
    def validate_target_amount(cls, v):
        if v <= 0:
            raise ValueError('Target amount must be greater than 0')
        return round(v, 2)


class MilestoneCreate(MilestoneBase):
    pass


class MilestoneUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=500)
    target_amount: Optional[float] = Field(None, gt=0)
    order_index: Optional[int] = None

    @validator('target_amount')
    def validate_target_amount(cls, v):
        if v is not None and v <= 0:
            raise ValueError('Target amount must be greater than 0')
        return round(v, 2) if v is not None else v


class MilestoneProgressUpdate(BaseModel):
    current_amount: float = Field(..., ge=0)

    @validator('current_amount')
    def validate_current_amount(cls, v):
        return round(v, 2)


class MilestoneResponse(MilestoneBase):
    id: int
    goal_id: int
    current_amount: float
    is_completed: bool
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# Update forward references
GoalWithMilestones.model_rebuild()

File: frontend/src/types/goal.ts
export type GoalType = 'SAVINGS' | 'DEBT_PAYOFF' | 'INVESTMENT' | 'EXPENSE_REDUCTION' | 'INCOME_INCREASE' | 'EMERGENCY_FUND';

export type GoalStatus = 'ACTIVE' | 'COMPLETED' | 'PAUSED' | 'CANCELLED';

export type GoalPriority = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export interface Goal {
  id: number;
  user_id: number;
  title: string;
  description?: string;
  goal_type: GoalType;
  priority: GoalPriority;
  status: GoalStatus;
  target_amount: number;
  current_amount: number;
  currency: string;
  progress_percentage: number;
  start_date: string;
  target_date?: string;
  is_milestone_based: boolean;
  created_at: string;
  updated_at?: string;
  milestones?: Milestone[];
}

export interface Milestone {
  id: number;
  goal_id: number;
  title: string;
  description?: string;
  target_amount: number;
  current_amount: number;
  is_completed: boolean;
  completed_at?: string;
  order_index: number;
  created_at: string;
  updated_at?: string;
}

export interface CreateGoalRequest {
  title: string;
  description?: string;
  goal_type: GoalType;
  priority: GoalPriority;
  target_amount: number;
  currency: string;
  target_date?: string;
  is_milestone_based: boolean;
}

export interface UpdateGoalRequest {
  title?: string;
  description?: string;
  goal_type?: GoalType;
  priority?: GoalPriority;
  target_amount?: number;
  currency?: string;
  target_date?: string;
  status?: GoalStatus;
  is_milestone_based?: boolean;
}

export interface GoalProgressUpdate {
  current_amount: number;
}

export interface CreateMilestoneRequest {
  title: string;
  description?: string;
  target_amount: number;
  order_index: number;
}

export interface UpdateMilestoneRequest {
  title?: string;
  description?: string;
  target_amount?: number;
  order_index?: number;
}

export interface MilestoneProgressUpdate {
  current_amount: number;
}

export interface GoalListResponse {
  items: Goal[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface GoalStatistics {
  total_goals: number;
  active_goals: number;
  completed_goals: number;
  total_target_amount: number;
  total_current_amount: number;
  overall_progress: number;
  goals_by_type: Record<string, number>;
  goals_by_priority: Record<string, number>;
}

export interface GoalFilters {
  status?: GoalStatus;
  goal_type?: GoalType;
  priority?: GoalPriority;
  search?: string;
}

// Helper types for UI
export interface GoalWithProgress extends Goal {
  daysRemaining?: number;
  isOverdue?: boolean;
  progressColor?: string;
}

export interface GoalFormData {
  title: string;
  description: string;
  goal_type: GoalType;
  priority: GoalPriority;
  target_amount: string;
  currency: string;
  target_date: string;
  is_milestone_based: boolean;
}

export interface MilestoneFormData {
  title: string;
  description: string;
  target_amount: string;
  order_index: number;
}

// Constants for UI
export const GOAL_TYPE_LABELS: Record<GoalType, string> = {
  SAVINGS: 'Накопления',
  DEBT_PAYOFF: 'Погашение долга',
  INVESTMENT: 'Инвестиции',
  EXPENSE_REDUCTION: 'Сокращение расходов',
  INCOME_INCREASE: 'Увеличение дохода',
  EMERGENCY_FUND: 'Резервный фонд'
};

export const GOAL_STATUS_LABELS: Record<GoalStatus, string> = {
  ACTIVE: 'Активная',
  COMPLETED: 'Завершена',
  PAUSED: 'Приостановлена',
  CANCELLED: 'Отменена'
};

export const GOAL_PRIORITY_LABELS: Record<GoalPriority, string> = {
  LOW: 'Низкий',
  MEDIUM: 'Средний',
  HIGH: 'Высокий',
  CRITICAL: 'Критический'
};

export const GOAL_PRIORITY_COLORS: Record<GoalPriority, string> = {
  LOW: 'text-gray-500',
  MEDIUM: 'text-blue-500',
  HIGH: 'text-orange-500',
  CRITICAL: 'text-red-500'
};

export const GOAL_STATUS_COLORS: Record<GoalStatus, string> = {
  ACTIVE: 'text-green-500',
  COMPLETED: 'text-blue-500',
  PAUSED: 'text-yellow-500',
  CANCELLED: 'text-gray-500'
};
